codes/c/chapter_hashing/hash_map_chaining.c

```diff
@@ -0,0 +1,235 @@
+    memset(newNode, 0, sizeof(Node));
+    newNode->key = key;
+    newNode->val = (char*)malloc(strlen(val) + 1);
+    strcpy(newNode->val,val);
+    newNode->val[strlen(val)] = '\0';
+    
+    Pair* pair = &hashmap->buckets[index];
+    Node* node = pair->node;
+    if(node == NULL) {
+       hashmap->buckets[index].node = newNode;
+       hashmap->size++;
+       return ;
+    }
+    while(node != NULL) {
+        if(node->key == key) {
+            // 释放先前分配的内存
+            free(node->val);
+            // 更新节点的值
+            node->val = (char*)malloc(strlen(val) + 1);
+            strcpy(node->val, val);
+            node->val[strlen(val)] = '\0';
+            return ;
+        }
+        if(node->next == NULL) {
+            break;
+        }
+        node = node->next;
+    }
+    node->next = newNode;
+    hashmap->size++;
+}
+
+/* 删除操作 */
+void removeItem(ArrayHashMap* hashmap, int key) {
+    int index = hashFunc(hashmap, key);
+    Pair* pair = &hashmap->buckets[index];
+    Node* node = pair->node;
+    // 保存后继的节点
+    Node* prev = NULL;
+    while(node != NULL) {
+        if(node->key == key) {
+            // 如果要删除的节点是桶的第一个节点
+            if(prev == NULL) {
+                pair->node = node->next;
+            } else {
+                prev->next = node->next;
+            }
+            // 释放内存
+            free(node->val);
+            free(node);
+            hashmap->size--;
+            return ;
+        }
+        prev = node;
+        node = node->next;
+    }
+    return ;
+}
+
+/* 扩容哈希表 */
+void extend(ArrayHashMap* hashmap) {
+    // 暂存原哈希表
+    Pair* oldBuckets = hashmap->buckets;
+    int oldCapacity = hashmap->capacity;
+
+    // 创建新的哈希表，重新分配一段空间
+    hashmap->capacity *= hashmap->extendRatio;
+    hashmap->buckets = (Pair*)malloc(sizeof(Pair) * hashmap->capacity);
+    memset(hashmap->buckets, 0, sizeof(Pair) * hashmap->capacity);
+    hashmap->size = 0;
+
+    // 将原哈希表中的键值对重新哈希到新的哈希表中
+    for(int i = 0; i < oldCapacity; i++) {
+        Node* node = oldBuckets[i].node;
+        while (node != NULL) {
+            put(hashmap, node->key, node->val);
+            node = node->next;
+        }
+    }
+
+    // 释放原哈希表的内存
+    for(int i = 0; i < oldCapacity; i++) {
+        Node* node = oldBuckets[i].node;
+        while (node != NULL) {
+            Node* temp = node;
+            node = node->next;
+            free(temp->val);
+            free(temp);
+        }
+    }
+    free(oldBuckets);
+}
+
+/* 打印哈希表 */
+void print(ArrayHashMap* hashmap) {
+    for(int i = 0; i < hashmap->capacity; i++) {
+        printf("[");
+        Pair* pair = &hashmap->buckets[i];
+        Node* node = pair->node;
+        while(node != NULL) {
+            if(node->val != NULL) {
+                printf("%d->%s, ",node->key,node->val);
+            }
+            node = node->next;
+        }
+        printf("]\n");
+    }
+    return ;
+}
+
+/* Driver Code */
+int main() {
+    /* 初始化哈希表 */
+    ArrayHashMap * map = InitArrayHashMap(4);
```

Comment:
`InitArrayHashMap(4)` --> `InitArrayHashMap()`
It is recommended to hide this parameter and write it in the `newStructName` function.